### ISSUE TITLE
Preserve x_forwaded_host on 301 redirect

### ISSUE THREAD

**Issue by kswitzer** [cloudfoundry/staticfile-buildpack#193]
What version of Cloud Foundry and CF CLI are you using? (i.e. What is the output of running `cf curl /v2/info && cf version`?
v1.4.0
cf version 6.50.0+4f0c3a2ce.2020-03-03

What version of the buildpack you are using?
v1.5.18

If you were attempting to accomplish a task, what was it you were attempting to do?
When leaving off the trailing slash in my browser on the URL with a directory it sends a 301 redirect and the forward location is set with the trailing slash but it changes the host Location URL to the internal CF URL instead of the CNAME/x-forwarded-host URL which may or may not be accessible to the end user.

curl -v https://www.example.com/directory
...
< HTTP/1.1 301 Moved Permanently
< Content-Length: 162
< Content-Type: text/html
...
< Location: http://example-internal.cf.example.com/directory/
< Server: nginx
...
<
<html>
<head><title>301 Moved Permanently</title></head>
<body>

What did you expect to happen?
I was expecting the x-forwarded-host to be preserved in the redirect Location in the 301 redirect with the addition of the trailing / (https://www.example.com/directory/)

What was the actual behavior?
The internal Host name was served in the 301 redirect (http://example-internal.cf.example.com/directory/) with the / appended.

Please confirm where necessary:
* [ ] I have included a log output
* [ ] My log includes an error message
* [ X] I have included steps for reproduction

**Comment by robdimsdale** [comment]
Hi @kswitzer, sorry for the delay in getting back to you. Is this still an issue for you? We weren't able to reproduce this with our CF installation.

**Comment by kswitzer** [comment]
@robdimsdale Yes, it is still an issue.  I was able to some what work around this issue by creating a custom nginx.conf, but seems as though it would be better if the x_forwarded_host was preserved on 301 redirect.  I am not sure if there is something unique to our environment that is causing this, but I can't think of anything off hand.  I do know that one of my colleagues had previously opened a similar issue relating to redirct to SSL and the x_forwarded_host was not being preserved by the static build pack.  I believe that issue was fixed ( Preserves x_forwarded_host on https #142 ).  Is there anything I can do to help you reproduce this issue?

**Comment by robdimsdale** [comment]
We don't observe the behavior you're describing. Here's some information about our test environment (specific hostnames have been changed but the behavior is the same):

```
❯ cf api
api endpoint:   https://api.internal-example.cf-app.com
api version:    2.169.0
```

```
❯ cf --version
cf version 6.53.0+8e2b70a4a.2020-10-01
```

```
❯ cf buildpacks
Getting buildpacks...

buildpack               position   enabled   locked   filename                                       stack
staticfile_buildpack    1          true      false    staticfile_buildpack-cflinuxfs3-v1.5.24.zip    cflinuxfs3
...
```

We have one app pushed using the Staticfile buildpack:

```
 ❯ cf apps
Getting apps in org system / space my-space as admin...
OK

name          requested state   instances   memory   disk   urls
website   started           1/1         1G       1G     website.internal-example.cf-app.com, website.external-example.com
```

This app has two routes bound to it - the internal route (the default when pushing) and the external route:

```
❯ cf routes
Getting routes for org system / space my-space as admin ...

space      host          domain                  port   path   type   apps          service
my-space   website   internal-example.cf-app.com                        website
my-space   website   external-example.com                              website
```


Here is what we see when we try to access a directory without the trailing slash (output edited for brevity):

```
❯ curl -v https://website.external-example.com/images
> GET /images HTTP/2
> Host: website.external-example.com
> Accept: */*
>
< HTTP/2 301
< content-type: text/html
< location: http://website.external-example.com/images/
```

The `location` is returned as expected - the external host rather than the internal host.

Some observations:
* We have different versions of the API, buildpacks, and CF cli.
* Our system is upgrading to HTTP/2
* We are initiating the connection over https (whereas your example initiates the connection over http).

Can you confirm a couple of things for us please:
1. The versions of the CF API and buildpacks (and the CF CLI, though that is unlikely to be relevant)
2. What routes your app is mapped to (e.g. `cf routes`).

**Comment by kswitzer** [comment]
There are a few more components in our setup, and not sure if you have those in your environment.  I think it might be possible to reproduce this by adding a CNAME DNS entry to your external route name.
Do you have the ability to add a CNAME DNS entry to your external route (website.external-example.com) and point your curl to that instead of the external route name and see if the CNAME is returned on 301 redirect or if the external route name is returned?

If the CNAME is not returned in the 301 redirect, but the original route name is returned then the issue has been reproduced.

**Comment by robdimsdale** [comment]
I do have a CNAME DNS entry to the external route and that's what I curled. I omitted that from the description above. I also omitted a redirect from HTTP to HTTPS.

**Comment by robdimsdale** [comment]
@kswitzer I think we're at the point where we can't debug this any further. I can't see anything in the buildpack configuration options that would cause this (or resolve it).

If you have additional thoughts or if you can reproduce this on an isolated cloudfoundry installation with isolated DNS configuration (i.e. separate from any load balancers, secondary DNS configuration, or custom CF route mapping) please feel free to reopen.

Sorry that we couldn't get to the bottom of this :confused:
